editor/editor_node.cpp

```diff
@@ -813,6 +814,11 @@ void EditorNode::_notification(int p_what) {
 			}
 #endif
 		} break;
+
+		case NOTIFICATION_PREDELETE: {
+			// Cannot be done in destructor since saving a resource emits EditorNode signals using memory that's being destroyed.
+			EditorHelp::save_script_doc_cache();
```

Comment:
I tried with a project with 3000 script files and it took a while to load the project. I was able to close the editor before all the script were parsed. Maybe it could be interesting to have an indicator that the script cache is fully loaded and check that indicator before saving the cache. Maybe even removing the cache file if not completely loaded when quitting the editor? I'm also wondering if it should be interesting to remove the cache file when a new script is modified preventing the cache to be outdated if the editor crash or something.

Reply:
Some thoughts!
- Addressing the first concern: I can make loading and saving use the worker thread, so they can't happen simultaneously. Or at least, I can make the save on Godot exit wait for the worker thread.
- If we delete the cache as soon as a script is modified, we force-recompile all scripts on basically every run of the editor since most folks edit a script at one point or another, so that doesn't feel viable.
- With your PR, we should be able to keep cache consistent by getting an update whenever a a file changes, so the docs are re-added, then saved on exit.
- I like the idea of deleting the cache file just before saving it, it safeguards against errors on editor shutdown.

Reply:
> Addressing the first concern: I can make loading and saving use the worker thread, so they can't happen simultaneously. Or at least, I can make the save on Godot exit wait for the worker thread.

I think you can keep it very simple by adding a global bool at false by default and to set it at true when the documentation for all scripts has been loaded or the cache is loaded. You skip writing the cache if the bool is not true on editor exit. That way, the cache will never contain partial documentation. I strongly suggest you don't use thread on exit, the thread will never finish.

> If we delete the cache as soon as a script is modified, we force-recompile all scripts on basically every run of the editor since most folks edit a script at one point or another, so that doesn't feel viable.

Not really because it's resaved on editor exit. The only time the documentation should be completely regenerated is if the editor crash. I think it's better to generate the doc then having a partial documentation. 

Personally I would try something like that:
- Global bool doc_scripts_ready = false
- Load cache or regenerate script doc
- Set doc_scripts_ready = true when done
- When new doc arrives, delete the cache on disk if exists (this cache is deprecated and should never be reused)
- On exit if doc_scripts_ready = true, write the cache on disk

That way:
- The cache on disk is never deprecated, worst case then doc needs to be regenerated on next startup
- It's simple, easy to implement and debug
- Manage crashes
- Manage exit before all the project is loaded
- When there's no cache you shoud always have a cache on disk that is up to date on the next startup

Reply:
Done!